### PR TITLE
Separate test infrastruture from tests.

### PR DESCRIPTION
**Pull request by stefanseefeld**
This patch moves the Numba test infrastructure out of its test suite into the `testing` module, thereby allowing other projects to use it with their own test suites.

### CODE REVIEW

**Comment by seibert** [comment]
This seems to be failing due to an argument conflict with existing unittest options?
```argparse.ArgumentError: argument -R/--refleak: conflicting option string(s): -R, --refleak```
(See the Travis-CI build for details)

**Comment by stefanseefeld** [comment]
Right. This turns out to be caused by us previously monkey-patching the unittest.main TestProgram to really become NumbaTestProgram, which created some weird circularity confusing the Python interpreter (so some functions would be executed twice).
I have removed that logic (so `unittest.main` isn't changed any longer), and everything appears to be working fine now.

File: numba/testing.py [hunk 1]
```diff
@@ -1,9 +1,522 @@
 from __future__ import print_function, division, absolute_import
 
 import sys
+import os
 import functools
-
+import numba.unittest_support as unittest
```
Comment (pitrou) [inline numba/testing.py]: As a style issue, the PEP 8 recommendation for imports is:
> Imports should be grouped in the following order:

> -    standard library imports
-  related third party imports
-  local application/library specific imports

> You should put a blank line between each group of imports. 

I like this rule as it makes dependencies much easier to figure out.

File: numba/testing.py [hunk 2]
```diff
@@ -1,9 +1,522 @@
...
+                load_tests = getattr(package, 'load_tests', None)
+                if load_tests:
+                    return load_tests(self, [], pattern)
+            except case.SkipTest as e:
+                return _make_skipped_test(name, e, self.suiteClass)
+            else:
+                return super(TestLoader, self).discover(start_dir, pattern, top)
+        else:
+            try:
+                __import__(start_dir)
```
Comment (pitrou) [inline numba/testing.py]: I suppose start_dir doesn't have slashes or dots at this point?

File: numba/testing.py [hunk 3]
```diff
@@ -1,9 +1,522 @@
...
+                module = sys.modules[start_dir]
+                load_tests = getattr(module, 'load_tests', None)
+                if load_tests:
+                    return load_tests(self, [], pattern)
+            finally:
+                return super(TestLoader, self).discover(start_dir, pattern, top)
```
Comment (pitrou) [inline numba/testing.py]: `return` in a `finally` is a bit weird. If an exception was raised, nothing will be returned regardless...

File: numba/testing.py [hunk 4]
```diff
@@ -1,9 +1,522 @@
...
+    from Queue import Empty as TimeoutError
+
+
+class TestLister(object):
+    """Simply list available tests rather than running them."""
+    def __init__(self):
+        pass
+
+    def run(self, test):
+        result = runner.TextTestResult(sys.stderr, True, 1)
```
Comment (pitrou) [inline numba/testing.py]: Just a nit, but `True` and `1` are not very explicit arguments if you don't name them ;-)

File: numba/testing.py [hunk 5]
```diff
@@ -1,9 +1,522 @@
...
+        parser.add_argument('-l', '--list', dest='list',
+                            action='store_true',
+                            help='List tests without running them')
+        parser.add_argument('--profile', dest='profile',
+                            action='store_true',
+                            help='Profile the test run')
+        return parser
+
+    def parseArgs(self, argv):
+        if '-l' in argv:
```
Comment (pitrou) [inline numba/testing.py]: You may add the PYVERSION guard here too.

File: numba/testing.py (hunk 2 above)
Comment (stefanseefeld) [inline numba/testing.py]: Hopefully not. But I have to admit that I struggled quite a bit trying to figure out the possible values of the inputs. (The docs are very sparse, and the code shows its semantic evolution, so it isn't quite clear what those variables are supposed (and allowed) to contain.)

File: numba/testing.py (hunk 1 above)
Comment (stefanseefeld) [inline numba/testing.py]: Thanks, corrected.

File: numba/testing.py (hunk 3 above)
Comment (stefanseefeld) [inline numba/testing.py]: Sorry, I'm not sure what I was thinking. I have replaced the 'finally:' by

  ```
  except:
    pass
  return ...
  ```

which should be clear and correct.

File: numba/testing.py (hunk 4 above)
Comment (stefanseefeld) [inline numba/testing.py]: This is another example where I didn't find any documentation. The implementation seemed to suggest that the second argument is checked as a boolean, while the third is eventually compared against '1' (verbosity).

Comment (pitrou) [inline numba/testing.py]: I was simply suggesting using named arguments ;-)

File: numba/testing.py (hunk 5 above)
Comment (stefanseefeld) [inline numba/testing.py]: I'm not sure I understand the purpose. There isn't anything that's available in recent (or even future) Python versions, or is there ?

File: numba/testing.py (hunk 4 above)
Comment (stefanseefeld) [inline numba/testing.py]: Oh, hah, yeah, I can certainly do that. Good suggestion ! :-)

File: numba/testing.py (hunk 5 above)
Comment (pitrou) [inline numba/testing.py]: Perhaps edge cases such as passing `-l` as parameter of an other option, but that's probably not useful. More importantly, actually testing the argparse argument in 3.4+ :-)

File: numba/testing.py (hunk 4 above)
Comment (seibert) [inline numba/testing.py]: I can merge this once you make this change.

Comment (stefanseefeld) [inline numba/testing.py]: Done.

**Pull request merged**